enhancements/ingress/haproxy-thread-tuning.md

```diff
@@ -0,0 +1,190 @@
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This proposal is to allow the cluster administrator to configure the number of
+connection handling threads within router pods.
+
+## Motivation
+
+As the number of routes and the volume of traffic through a cluster increases,
+eventually the router pod will reach a limit to how many connections it can
+handle. In order to increase the maximum number of connections, administrators
+can split the routes into multiple shards, each shard being handled by a
+separate router.  More router pods can also be deployed without sharding,
+allowing more connections to be handled, but this comes with more overhead, as
+each router must separately verify the health of its backends.  These redundant
+health checks increase the volume of traffic passing through the cluster, and
+can cancel out much of the performance gain from deploying more router pods.
+
+In OpenShift 3.x, users could configure routers to increase the number of
+threads that HAProxy spawns, allowing a single router to handle more
+connections without additional health check overhead. This document proposes to
+allow users to employ a similar strategy, and provide a field within the
+IngressController API to specify the number of threads allocated within router
+pods
+
+### Goals
+
+Provide an API for configuring the number of threads handling connections in
+router pods
+
+### Non-Goals
+
+Expose additional performance tuning parameters available within HAProxy
+
+## Proposal
+
+Add the subresource `threading` to the IngressController API. It will currently
+contain one field, `count`.
+
+```go
+type IngressControllerSpec struct {
+	// ...
+	// Existing fields
+	// ...
+
+	// threading defines parameters for configuring threading options within
+	// routers created under this IngressController. See specific threading
+	// fields for their respective definitions and default values.
+	//
+	// +optional
+	Threading IngressControllerThreading `json:"threading,omitempty"`
+}
+
+type IngressControllerThreading struct {
+	// count defines the number of threads created per router pod. Creating
+	// more threads allows each router pod to handle more connections, at the
+	// cost of more system resources used. If this field is empty, the
+	// IngressController will use a default value of 4 threads.
+	//
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Minimum=4
+	// +optional
+	Count int `json:"count,omitempty"`
+}
+```
+
+When unset, the ingress operator will provision routers with 4 threads,
+matching the existing behavior in OpenShift 4.7.
+
+It would be possible to implement this field as a simple integer within the
+`IngressControllerSpec` as `spec.threadCount` or something similar, however
+there are other tunable threading performance options that HAProxy exposes, and
```

Comment:
Perhaps cite "cpu pinning" as a concrete example of something we may want to consider in the future.